### ISSUE TITLE
34 alive mutations on rom-sql

### ISSUE THREAD

**Issue by paneq** [RailsEventStore/rails_event_store#301]

**Comment by paneq** [comment]
I looked into a mutation changing `(savepoint: true)` to `(savepoint: false)` in:

```ruby
        def create(serialized_records, stream: nil, expected_version: nil)
          events.transaction(savepoint: true) do
            events.changeset(CreateEventsChangeset, serialized_records).commit.tap do
              if stream
                link(
                  serialized_records.map(&:event_id),
                  stream,
                  expected_version || ExpectedVersion.any,
                  global_stream: true
                )
              end
            end
          end
        end
```

and the  I don't know why the test does not kill the mutation.

```ruby
    specify "nested transaction - events still not persisted if append failed" do
      repository = EventRepository.new
      repository.append_to_stream([
        event = SRecord.new(event_id: SecureRandom.uuid),
      ], default_stream, RubyEventStore::ExpectedVersion.none)

      rom_db.transaction do
        expect do
          repository.append_to_stream([
            SRecord.new(
              event_id: '9bedf448-e4d0-41a3-a8cd-f94aec7aa763'
            ),
          ], default_stream, RubyEventStore::ExpectedVersion.none)
        end.to raise_error(RubyEventStore::WrongExpectedEventVersion)
        expect(repository.has_event?('9bedf448-e4d0-41a3-a8cd-f94aec7aa763')).to be_falsey
        expect(repository.read_all_streams_forward(:head, 2)).to eq([event])
      end
      expect(repository.has_event?('9bedf448-e4d0-41a3-a8cd-f94aec7aa763')).to be_falsey
      expect(repository.read_all_streams_forward(:head, 2)).to eq([event])
    end
```

logs (when false):

```
I, [2018-04-17T18:43:31.098243 #10802]  INFO -- : (0.000028s) BEGIN
I, [2018-04-17T18:43:31.099967 #10802]  INFO -- : (0.000044s) INSERT INTO `event_store_events` (`id`, `event_type`, `metadata`, `data`, `created_at`) VALUES ('9bedf448-e4d0-41a3-a8cd-f94aec7aa763', 'ce40960d-52e0-4ce8-a220-87aa156cccfa', 'c8335ad6-dd35-4ea7-a7a2-ffdf89b74779', 'c8add981-9923-4a19-86dc-cb2aa3318ab9', '2018-04-17 16:43:31.099786')
I, [2018-04-17T18:43:31.100323 #10802]  INFO -- : (0.000060s) SELECT `event_store_events`.`id`, `event_store_events`.`event_type`, `event_store_events`.`metadata`, `event_store_events`.`data`, `event_store_events`.`created_at` FROM `event_store_events` WHERE (`id` IN (2)) ORDER BY `event_store_events`.`id`
I, [2018-04-17T18:43:31.100647 #10802]  INFO -- : (0.000039s) SELECT `event_store_events`.`id` FROM `event_store_events` WHERE (`id` IN ('9bedf448-e4d0-41a3-a8cd-f94aec7aa763')) ORDER BY `event_store_events`.`id`
E, [2018-04-17T18:43:31.101101 #10802] ERROR -- : SQLite3::ConstraintException: UNIQUE constraint failed: event_store_events_in_streams.stream, event_store_events_in_streams.position: INSERT INTO `event_store_events_in_streams` (`stream`, `position`, `event_id`, `created_at`) VALUES ('stream', 0, '9bedf448-e4d0-41a3-a8cd-f94aec7aa763', '2018-04-17 16:43:31.100941')
I, [2018-04-17T18:43:31.101694 #10802]  INFO -- : (0.000054s) SELECT count(*) AS 'count' FROM (SELECT `event_store_events`.`id`, `event_store_events`.`event_type`, `event_store_events`.`metadata`, `event_store_events`.`data`, `event_store_events`.`created_at` FROM `event_store_events` WHERE (`id` = '9bedf448-e4d0-41a3-a8cd-f94aec7aa763') ORDER BY `event_store_events`.`id` LIMIT 1) AS 't1' LIMIT 1
I, [2018-04-17T18:43:31.120844 #10802]  INFO -- : (0.000050s) ROLLBACK
```

because the test is failing...

**Comment by paneq** [comment]
aaaah that's because there is another, more specific group of tests which does not kill the mutation:

```
module RubyEventStore::ROM::Repositories
  RSpec.describe Events do
```

@joelvh https://github.com/mbj/mutant#test-selection

> Mutant uses the "longest rspec example group descriptions prefix match" to select the tests to run.
> Example for a subject like Foo::Bar#baz it will run all example groups with description prefixes in Foo::Bar#baz, Foo::Bar and Foo. The order is important, so if mutant finds example groups in the current prefix level, these example groups must kill the mutation.

@joelvh I suggest treating Repositories::* as black-box and testing on the level of EventRepository

**Comment by joelvh** [comment]
@paneq Yes, I agree - I ended up creating more tests to get Mutant to shut up, but it kept complaining. You can remove those and see what happens. Let me know how you'd like me to help so we don't step on toes.

**Comment by paneq** [comment]
> I ended up creating more tests to get Mutant to shut up, but it kept complaining. You can remove those and see what happens.

I don't see a reason to remove them if you find them helpful. Just change their subject to `RubyEventStore::ROM::EventRepository` so they can be used for killing mutation as well.

**Comment by joelvh** [comment]
I’ve removed them and will probably bring them back in the refactor branch that will set the stage for more modularity in ROM.

**Comment by joelvh** [comment]
Tests for internals removed in #304

**Comment by joelvh** [comment]
Mutations addressed in #314
